pkg/networkservice/common/connect/server.go

```diff
@@ -95,60 +120,97 @@ func (s *connectServer) Close(ctx context.Context, conn *networkservice.Connecti
+		// For some reason we have changed the clientURL, so we need to close and delete the existing client.
 		if _, clientErr := connInfo.client.Close(ctx, conn); clientErr != nil {
 			logEntry.Warnf("failed to close client: %+v", clientErr)
 		}
+
+		if onClose, ok := s.clientsCloseFuncs[connInfo.client]; ok {
+			onClose()
+		} else {
+			logEntry.Warnf("failed to delete client connected to %s", connInfo.clientURL.String())
+		}
 	}
 
-	// Fast path if we already have client for the clientURL, use it.
+	// Fast path if we already have client for the clientURL and we chould not reconnect, use it.
 	client, loaded := s.clients.Load(clientURL.String())
+	var onClose closeFunc
 	if !loaded {
 		// If not, create and LoadOrStore a new one.
-		newClient, cancel := s.newClient(clientURL)
-		client, loaded = s.clients.LoadOrStore(clientURL.String(), newClient)
-		if loaded {
-			// No one will use `newClient`, it should be canceled.
-			cancel()
-		}
+		client, onClose = s.newClient(clientURL)
+		s.clients.Store(clientURL.String(), client)
+		s.clientsCloseFuncs[client] = onClose
 	}
 
-	s.connInfos.Store(conn.GetId(), connectionInfo{
-		clientURL: clientURL,
-		client:    client,
-	})
-
-	return client
+	return client, onClose
 }
 
-func (s *connectServer) newClient(clientURL *url.URL) (networkservice.NetworkServiceClient, context.CancelFunc) {
+// newClient has to be called under clientsMutex protection
+func (s *connectServer) newClient(clientURL *url.URL) (networkservice.NetworkServiceClient, closeFunc) {
 	clientPtr := new(networkservice.NetworkServiceClient)
 
 	ctx, cancel := context.WithCancel(s.ctx)
-	onClose := func() {
+	onClose := func() bool {
+		client, ok := s.clients.Load(clientURL.String())
+		if ok {
+			s.clients.Delete(clientURL.String())
+		}
+
+		if client != *clientPtr {
+			cancel()
+			delete(s.clientsCloseFuncs, *clientPtr)
+			return true
+		}
+
 		if deleted := s.clients.Delete(clientURL.String()); deleted {
 			cancel()
+			delete(s.clientsCloseFuncs, *clientPtr)
+			return true
 		}
+
+		return false
+	}
+
+	*clientPtr = clienturl.NewClient(clienturlctx.WithClientURL(ctx, clientURL), s.clientFactory, s.clientDialOptions...)
+
+	return *clientPtr, onClose
+}
+
+func (s *connectServer) closeClient(client networkservice.NetworkServiceClient) {
+	s.clientsMutex.Lock()
+	defer s.clientsMutex.Unlock()
+
+	if onClose, ok := s.clientsCloseFuncs[client]; ok {
+		onClose()
 	}
+}
 
-	*clientPtr = chain.NewNetworkServiceClient(
-		&connectClient{onClose: onClose},
-		clienturl.NewClient(clienturlctx.WithClientURL(ctx, clientURL), s.clientFactory, s.clientDialOptions...),
-	)
+func (s *connectServer) deleteClient(client networkservice.NetworkServiceClient) {
+	s.clientsMutex.Lock()
+	defer s.clientsMutex.Unlock()
 
-	return *clientPtr, cancel
+	if onClose, ok := s.clientsCloseFuncs[client]; ok {
+		for {
+			if onClose() {
+				return
+			}
+		}
```

Comment:
Why we cannot try to call close one time?

Reply:
onClose hase to be called while exist at least one ref.

Reply:
What if we have 1e3 clients? Do we really need in this case to call on close 1e3 times?

Reply:
Done